### DISCUSSION TITLE
Application management

### DISCUSSION THREAD

**Discussion by DamnedElric** [netbox-community/netbox#10985] [General]
The data in Netbox is incredibly useful, both for automation and as documentation. But when it comes to the documentation use case, I feel like it's missing an Application component. In my view, this is a higher level component than a service. 

Let me make things a bit more concrete with a contrived example: 

Organization has a Web Shop, which is served by 3 web servers behind a load balancer, which talks to a database server which in turn is replicated to another datacenter. One of the web servers is physical, the other two are virtual.

If the Web Shop is down, and the new guy is on call, he will have a hard time figuring all the possible places where the problem could be. But if we could describe the Web Shop application in Netbox, it would be much easier to start poking around in the relevant places. Maybe the switch connecting a Web Server to the database is borked? 

We can sort of do this using Tags at the moment, but I suspect a dedicated Application component would be cleaner. Might even make it possible to visualise an Application and all of its dependencies as a graph. 

Any thoughts on this? Thanks!

**Comment by richbibby** [comment]
https://docs.netbox.dev/en/stable/models/ipam/service/

**Reply by candlerb** [reply]
To me, tags sounds like exactly the right approach: "this application is built up of components X, Y and Z" (which could be anything: devices, virtual machines, circuits etc).

For a full graph you need directed dependency links: "this component (X) depends on component (Y)".  You could write a plugin for that.  But I don't think it helps all that much, except perhaps for some trivial alert suppression scenarios: e.g. silence alerts for all devices in site A, if site A's circuit has gone down.

I try to do the opposite, as described in [this philosophy on alerting](https://docs.google.com/document/d/199PqyG3UsyXlwieHaqbGiWVa8eMWi8zzAn0YfcApr8Q/edit):
* Focus on customer-affecting problems ("symptom" checks), e.g. "website is rejecting credit card payments"
* Send out alerts for those
* Within the alert body, include all system-level ("cause" checks) which are firing at the same time, e.g. "link to site A is down, mysql database is down"

Rather than suppressing child alerts, you're enriching parent alerts with child data.

**Comment by mtinberg** [comment]
I'm not sure Netbox is quite the right tool for that, although someone may write a plugin for this, Netbox is more focused on the physical infrastructure and providing info for config management on top but what you are describing is an ITIL CMDB or monitoring systems that is documenting complex service topologies.  Even without a specialized CMDB tool to model the service dependencies, hand-making functional diagrams using mermaid, graphviz, xfig, Visio or whatever can help speed this kind of troubleshooting.


—

Mark Tinberg ***@***.***>

Division of Information Technology-Network Services

University of Wisconsin-Madison


________________________________
From: Elric ***@***.***>
Sent: Saturday, November 19, 2022 4:31 PM
To: netbox-community/netbox ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [netbox-community/netbox] Application management (Discussion #10985)


The data in Netbox is incredibly useful, both for automation and as documentation. But when it comes to the documentation use case, I feel like it's missing an Application component. In my view, this is a higher level component than a service.

Let me make things a bit more concrete with a contrived example:

Organization has a Web Shop, which is served by 3 web servers behind a load balancer, which talks to a database server which in turn is replicated to another datacenter. One of the web servers is physical, the other two are virtual.

If the Web Shop is down, and the new guy is on call, he will have a hard time figuring all the possible places where the problem could be. But if we could describe the Web Shop application in Netbox, it would be much easier to start poking around in the relevant places. Maybe the switch connecting a Web Server to the database is borked?

We can sort of do this using Tags at the moment, but I suspect a dedicated Application component would be cleaner. Might even make it possible to visualise an Application and all of its dependencies as a graph.

Any thoughts on this? Thanks!
